### PR TITLE
Theia extension for SSH key management

### PR DESCRIPTION
**Pull request by azatsarynnyy**
Signed-off-by: Artem Zatsarynnyi <azatsary@redhat.com>

CQs:
- [x] [Theia 0.3.7](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=15845)
- [x] [axios 0.17.1](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=15800)
- [x] [lerna 2.4.0](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=15841)

Related issues: https://github.com/eclipse/che/issues/8727, https://github.com/eclipse/che/issues/8738, https://github.com/eclipse/che/issues/8737.

Unit tests are yet not implemented, probably will use [Mocha](https://mochajs.org/) for those purposes.

#### Backend
The pull request adds a server-side support for SSH keys related API.  There is server-side implementation of `SshKeyServer`, it plays the role of a JSON-RPC endpoint for client requests. There is also a `SshKeyManager` that may be used by server-side components interactions with SSH key API.

The implementation uses specific REST client, which wraps [axios](https://github.com/axios/axios) HTTP client, in order to reach out the SSH service while the service itself is located on the workspace master instance. The workspace master API endpoint URL is fetched out of the environment variable `CHE_API`, if it is not set the REST client initialization fails.

#### Frontend
The pull request adds the following commands to Commands Palette (shortcut `F1`):
SSH commands

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                "${workspaceRoot}/electron-app/src-gen/backend/main.js",
+                "${workspaceRoot}/electron-app/lib/**/*.js",
+                "${workspaceRoot}/node_modules/@theia/*/lib/**/*.js"
+            ],
+            "smartStep": true,
+            "internalConsoleOptions": "openOnSessionStart",
+            "outputCapture": "std"
+        }
+    ]
+}
```
Comment (vinokurig) [inline .vscode/launch.json]: Could you please add new lines

**Comment by azatsarynnyy** [comment]
Added one more commit to propose to download the generated public key

File: .vscode/launch.json (hunk 1 above)
Comment (azatsarynnyy) [reply .vscode/launch.json]: done

**Review by vparfonov — APPROVED** [review]

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,60 @@
+{
+    // Use IntelliSense to learn about possible Node.js debug attributes.
```
Comment (benoitf) [inline .vscode/launch.json:2]: AFAIK json format can't contain any comments

File: README.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    cd browser-app
+    yarn watch
+
+Launch `Start Browser Backend` configuration from VS code.
+
+Open http://localhost:3000 in the browser.
+
+## Developing with the Electron example
+
+Start watching of the hello world extension.
```
Comment (benoitf) [inline README.md]: it's not hello world so maybe readme needs to be updated/reduced

File: browser-app/package.json [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    "@theia/markers": "latest",
+    "@theia/messages": "latest",
+    "@theia/monaco": "latest",
+    "@theia/navigator": "latest",
+    "@theia/preferences": "latest",
+    "@theia/process": "latest",
+    "@theia/terminal": "latest",
+    "@theia/typescript": "latest",
+    "@theia/workspace": "latest",
+    "che-theia-ssh-extension": "0.0.0"
```
Comment (benoitf) [inline browser-app/package.json]: I see only theia/core as a CQ
is browser-app dependencies are only build and test dependencies then ?
ot we should remove browser-app module ?

File: package.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
+{
+  "private": true,
+  "scripts": {
+    "prepare": "lerna run prepare",
```
Comment (benoitf) [inline package.json:4]: are these scripts covered as well by build and test dependencies CQs ?

File: .vscode/launch.json (hunk 2 above)
Comment (azatsarynnyy) [reply .vscode/launch.json:2]: @benoitf you're right. JSON specification doesn't allow comments.
Guys from the VSCode team added a separate language mode into VSCode - `jsonc` for _JSON with Comments_. All VSCode-owned files (launch.json, tasks.json, keybindings.json etc.) are associated with that mode in VSCode.
Also, they created [jsonc-parser node module](https://www.npmjs.com/package/jsonc-parser) which takes into account the comments.

File: package.json (hunk 5 above)
Comment (azatsarynnyy) [reply package.json:4]: @benoitf thanks for pointing this out.
lerna isn't covered. I've created a [CQ](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=15841).

File: browser-app/package.json (hunk 4 above)
Comment (azatsarynnyy) [reply browser-app/package.json]: I've created CQ for the entire Theia 0.3.7
https://dev.eclipse.org/ipzilla/show_bug.cgi?id=15845

File: ssh/package.json [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    "lib",
+    "src"
+  ],
+  "dependencies": {
+    "@theia/core": "0.3.7",
+    "axios": "0.17.1"
+  },
+  "devDependencies": {
+    "rimraf": "2.6.2",
+    "typescript": "latest"
```
Comment (benoitf) [inline ssh/package.json]: here version should be specified

File: ssh/package.json [hunk 7]
```diff
@@ -0,0 +1,49 @@
+{
+  "name": "che-theia-ssh-extension",
```
Comment (AndrienkoAleksandr) [inline ssh/package.json]: @benoitf  @azatsarynnyy What package scope should we use for release extension? Or we don't want to use it? Theia for example use package scope `"theia"`, so Theia npm packages for example calls with this package scope: `"@theia/core"`, `"@theia/workspace"` and so on. And by this package scope we can quickly filter on the npm registry all packages for Theia. And package scope is useful if want use local npm registry (for example we want use local npm registry for internal testing before public release new extension).

File: README.md (hunk 3 above)
Comment (azatsarynnyy) [reply README.md]: I've updated README.md and contributed [a related fix](https://github.com/theia-ide/generator-theia-extension/pull/15) to Theia extension generator. Thanks for noticing!

File: ssh/package.json (hunk 6 above)
Comment (azatsarynnyy) [reply ssh/package.json]: done

File: ssh/package.json (hunk 7 above)
Comment (azatsarynnyy) [reply ssh/package.json]: @AndrienkoAleksandr I have no idea what package scope should we use for our extensions.

**Comment by azatsarynnyy** [comment]
Unit tests will be committed in separate PRs

**Comment by dkulieshov** [comment]
Created a separate [pull request](https://github.com/eclipse/che-theia-ssh-plugin/pull/4) for tests.

**Pull request merged**